File: public/sdk/java/AuthClient.java
package com.authsystem.sdk;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;
import okhttp3.*;
import oshi.SystemInfo;
import oshi.hardware.ComputerSystem;
import oshi.hardware.HardwareAbstractionLayer;
import oshi.hardware.NetworkIF;
import oshi.hardware.platform.mac.MacComputerSystem;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.List;
import java.util.concurrent.*;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Professional-grade Java Authentication SDK with HWID-based Device Binding
 * Cross-platform support using OSHI library
 */
public class AuthClient {
    
    private static final Logger LOGGER = Logger.getLogger(AuthClient.class.getName());
    private static final long HEARTBEAT_INTERVAL = 5 * 60 * 1000; // 5 minutes
    
    private final String apiBaseUrl;
    private final String apiKey;
    private final OkHttpClient httpClient;
    private final Gson gson;
    private final HWIDGenerator hwidGenerator;
    
    private String jwtToken;
    private ScheduledExecutorService heartbeatExecutor;
    private String cachedHwid;
    
    /**
     * Private constructor - use Builder pattern
     */
    private AuthClient(Builder builder) {
        this.apiBaseUrl = builder.apiBaseUrl.replaceAll("/$", "");
        this.apiKey = builder.apiKey;
        this.httpClient = builder.httpClient != null ? builder.httpClient : createDefaultHttpClient();
        this.gson = new GsonBuilder().create();
        this.hwidGenerator = new HWIDGenerator();
    }
    
    /**
     * Get the current device HWID (cached)
     */
    public String getHWID() {
        if (cachedHwid == null) {
            cachedHwid = hwidGenerator.generate();
        }
        return cachedHwid;
    }
    
    /**
     * Register a new user account
     */
    public CompletableFuture<AuthResponse> register(
            String username,
            String password,
            String email,
            String licenseKey
    ) {
        RegisterRequest request = new RegisterRequest();
        request.apiKey = this.apiKey;
        request.username = username;
        request.password = password;
        request.email = email;
        request.licenseKey = licenseKey;
        request.hwid = getHWID();
        
        return sendRequest("POST", "/api/user/create", request, AuthResponse.class);
    }
    
    /**
     * Login with username and password
     * The server will automatically bind this device's HWID to the account
     */
    public CompletableFuture<LoginResponse> login(String username, String password) {
        LoginRequest request = new LoginRequest();
        request.apiKey = this.apiKey;
        request.username = username;
        request.password = password;
        request.hwid = getHWID();
        
        return sendRequest("POST", "/api/user/login", request, LoginResponse.class)
                .thenApply(response -> {
                    if (response.success && response.token != null) {
                        this.jwtToken = response.token;
                        startHeartbeat();
                    }
                    return response;
                });
    }
    
    /**
     * Validate a license key
     */
    public CompletableFuture<ValidationResponse> validateLicense(String licenseKey) {
        ValidateRequest request = new ValidateRequest();
        request.apiKey = this.apiKey;
        request.licenseKey = licenseKey;
        request.hwid = getHWID();
        
        return sendRequest("POST", "/api/auth/validate", request, ValidationResponse.class);
    }
    
    /**
     * Check if the current session is still valid
     */
    public CompletableFuture<SessionResponse> checkSession() {
        return sendRequest("GET", "/api/user/session", null, SessionResponse.class);
    }
    
    /**
     * Logout and invalidate the current session
     */
    public CompletableFuture<AuthResponse> logout() {
        stopHeartbeat();
        return sendRequest("POST", "/api/user/logout", null, AuthResponse.class)
                .thenApply(response -> {
                    this.jwtToken = null;
                    return response;
                });
    }
    
    /**
     * Start the session heartbeat
     */
    private void startHeartbeat() {
        stopHeartbeat();
        heartbeatExecutor = Executors.newSingleThreadScheduledExecutor();
        heartbeatExecutor.scheduleAtFixedRate(
                this::runHeartbeat,
                HEARTBEAT_INTERVAL,
                HEARTBEAT_INTERVAL,
                TimeUnit.MILLISECONDS
        );
    }
    
    /**
     * Stop the session heartbeat
     */
    private void stopHeartbeat() {
        if (heartbeatExecutor != null) {
            heartbeatExecutor.shutdown();
            heartbeatExecutor = null;
        }
    }
    
    /**
     * Run a single heartbeat check
     */
    private void runHeartbeat() {
        checkSession()
                .thenAccept(response -> LOGGER.fine("Heartbeat successful"))
                .exceptionally(throwable -> {
                    LOGGER.warning("Heartbeat failed: " + throwable.getMessage());
                    return null;
                });
    }
    
    /**
     * Send HTTP request to the API
     */
    private <T> CompletableFuture<T> sendRequest(
            String method,
            String endpoint,
            Object payload,
            Class<T> responseClass
    ) {
        CompletableFuture<T> future = new CompletableFuture<>();
        
        try {
            String url = apiBaseUrl + endpoint;
            Request.Builder requestBuilder = new Request.Builder().url(url);
            
            // Add HWID header
            requestBuilder.addHeader("X-HWID", getHWID());
            
            // Add JWT token if available
            if (jwtToken != null && !jwtToken.isEmpty()) {
                requestBuilder.addHeader("Authorization", "Bearer " + jwtToken);
            }
            
            // Add payload if present
            if (payload != null) {
                String json = gson.toJson(payload);
                RequestBody body = RequestBody.create(
                        json,
                        MediaType.get("application/json; charset=utf-8")
                );
                requestBuilder.method(method, body);
            } else {
                requestBuilder.method(method, null);
            }
            
            Request request = requestBuilder.build();
            
            httpClient.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    future.completeExceptionally(
                            new AuthException("Network request failed", "NETWORK_ERROR", e)
                    );
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        String responseBody = response.body().string();
                        
                        if (!response.isSuccessful()) {
                            ErrorResponse error = gson.fromJson(responseBody, ErrorResponse.class);
                            future.completeExceptionally(
                                    new AuthException(
                                            error.message != null ? error.message : "Request failed",
                                            error.errorCode != null ? error.errorCode : "REQUEST_FAILED"
                                    )
                            );
                        } else {
                            T result = gson.fromJson(responseBody, responseClass);
                            future.complete(result);
                        }
                    } catch (Exception e) {
                        future.completeExceptionally(
                                new AuthException("Failed to parse response", "PARSE_ERROR", e)
                        );
                    } finally {
                        response.close();
                    }
                }
            });
            
        } catch (Exception e) {
            future.completeExceptionally(
                    new AuthException("Failed to create request", "REQUEST_ERROR", e)
            );
        }
        
        return future;
    }
    
    /**
     * Create default HTTP client with proper configuration
     */
    private OkHttpClient createDefaultHttpClient() {
        return new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .addInterceptor(new LoggingInterceptor())
                .build();
    }
    
    /**
     * Cleanup resources
     */
    public void close() {
        stopHeartbeat();
        if (httpClient != null) {
            httpClient.dispatcher().executorService().shutdown();
            httpClient.connectionPool().evictAll();
        }
    }
    
    // ==================== Builder Pattern ====================
    
    public static class Builder {
        private String apiBaseUrl;
        private String apiKey;
        private OkHttpClient httpClient;
        
        public Builder apiBaseUrl(String apiBaseUrl) {
            this.apiBaseUrl = apiBaseUrl;
            return this;
        }
        
        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            return this;
        }
        
        public Builder httpClient(OkHttpClient httpClient) {
            this.httpClient = httpClient;
            return this;
        }
        
        public AuthClient build() {
            if (apiBaseUrl == null || apiBaseUrl.isEmpty()) {
                throw new IllegalArgumentException("API base URL is required");
            }
            if (apiKey == null || apiKey.isEmpty()) {
                throw new IllegalArgumentException("API key is required");
            }
            return new AuthClient(this);
        }
    }
    
    // ==================== HWID Generator ====================
    
    /**
     * Cross-platform Hardware ID generator using OSHI
     */
    private static class HWIDGenerator {
        
        private final SystemInfo systemInfo;
        
        public HWIDGenerator() {
            this.systemInfo = new SystemInfo();
        }
        
        /**
         * Generate unique HWID using multiple hardware identifiers
         */
        public String generate() {
            try {
                HardwareAbstractionLayer hardware = systemInfo.getHardware();
                StringBuilder hwidBuilder = new StringBuilder();
                
                // Primary: System UUID and Serial
                try {
                    ComputerSystem computerSystem = hardware.getComputerSystem();
                    String uuid = computerSystem.getHardwareUUID();
                    String serial = computerSystem.getSerialNumber();
                    
                    if (uuid != null && !uuid.isEmpty()) {
                        hwidBuilder.append(uuid).append("-");
                    }
                    if (serial != null && !serial.isEmpty()) {
                        hwidBuilder.append(serial).append("-");
                    }
                } catch (Exception e) {
                    LOGGER.log(Level.WARNING, "Failed to get system UUID", e);
                }
                
                // Secondary: Processor ID
                try {
                    String processorId = hardware.getProcessor().getProcessorIdentifier().getProcessorID();
                    if (processorId != null && !processorId.isEmpty()) {
                        hwidBuilder.append(processorId).append("-");
                    }
                } catch (Exception e) {
                    LOGGER.log(Level.WARNING, "Failed to get processor ID", e);
                }
                
                // Tertiary: MAC Address
                try {
                    List<NetworkIF> networkIFs = hardware.getNetworkIFs();
                    for (NetworkIF net : networkIFs) {
                        if (net.getMacaddr() != null && !net.getMacaddr().isEmpty() &&
                                !net.getMacaddr().equals("00:00:00:00:00:00")) {
                            hwidBuilder.append(net.getMacaddr());
                            break;
                        }
                    }
                } catch (Exception e) {
                    LOGGER.log(Level.WARNING, "Failed to get MAC address", e);
                }
                
                // Fallback
                if (hwidBuilder.length() == 0) {
                    String fallback = System.getProperty("user.name") + "-" +
                            System.getProperty("os.name") + "-" +
                            System.getProperty("os.arch");
                    hwidBuilder.append(fallback);
                }
                
                return hashString(hwidBuilder.toString());
                
            } catch (Exception e) {
                LOGGER.log(Level.SEVERE, "HWID generation failed", e);
                // Ultimate fallback
                return hashString(System.getProperty("user.name") + "-" + System.currentTimeMillis());
            }
        }
        
        /**
         * Hash a string using SHA-256
         */
        private String hashString(String input) {
            try {
                MessageDigest digest = MessageDigest.getInstance("SHA-256");
                byte[] hash = digest.digest(input.getBytes(StandardCharsets.UTF_8));
                
                StringBuilder hexString = new StringBuilder();
                for (byte b : hash) {
                    String hex = Integer.toHexString(0xff & b);
                    if (hex.length() == 1) hexString.append('0');
                    hexString.append(hex);
                }
                return hexString.toString();
                
            } catch (NoSuchAlgorithmException e) {
                // Fallback to Base64 if SHA-256 not available
                return Base64.getEncoder().encodeToString(input.getBytes(StandardCharsets.UTF_8));
            }
        }
    }
    
    // ==================== Request Models ====================
    
    private static class RegisterRequest {
        String apiKey;
        String username;
        String password;
        String email;
        String licenseKey;
        String hwid;
    }
    
    private static class LoginRequest {
        String apiKey;
        String username;
        String password;
        String hwid;
    }
    
    private static class ValidateRequest {
        String apiKey;
        String licenseKey;
        String hwid;
    }
    
    // ==================== Response Models ====================
    
    public static class AuthResponse {
        @SerializedName("success")
        public boolean success;
        
        @SerializedName("message")
        public String message;
        
        @SerializedName("userId")
        public String userId;
    }
    
    public static class LoginResponse {
        @SerializedName("success")
        public boolean success;
        
        @SerializedName("message")
        public String message;
        
        @SerializedName("token")
        public String token;
        
        @SerializedName("user")
        public UserInfo user;
    }
    
    public static class UserInfo {
        @SerializedName("id")
        public String id;
        
        @SerializedName("username")
        public String username;
        
        @SerializedName("email")
        public String email;
        
        @SerializedName("hwid")
        public String hwid;
        
        @SerializedName("hwidLocked")
        public boolean hwidLocked;
        
        @SerializedName("ip")
        public String ip;
        
        @SerializedName("isFirstLogin")
        public boolean isFirstLogin;
        
        @SerializedName("expiresAt")
        public String expiresAt;
    }
    
    public static class ValidationResponse {
        @SerializedName("success")
        public boolean success;
        
        @SerializedName("message")
        public String message;
        
        @SerializedName("valid")
        public boolean valid;
        
        @SerializedName("expiresAt")
        public String expiresAt;
    }
    
    public static class SessionResponse {
        @SerializedName("success")
        public boolean success;
        
        @SerializedName("message")
        public String message;
        
        @SerializedName("valid")
        public boolean valid;
        
        @SerializedName("user")
        public UserInfo user;
    }
    
    private static class ErrorResponse {
        @SerializedName("success")
        public boolean success;
        
        @SerializedName("message")
        public String message;
        
        @SerializedName("errorCode")
        public String errorCode;
    }
    
    // ==================== Custom Exception ====================
    
    public static class AuthException extends Exception {
        private final String errorCode;
        
        public AuthException(String message, String errorCode) {
            super(message);
            this.errorCode = errorCode;
        }
        
        public AuthException(String message, String errorCode, Throwable cause) {
            super(message, cause);
            this.errorCode = errorCode;
        }
        
        public String getErrorCode() {
            return errorCode;
        }
    }
    
    // ==================== Logging Interceptor ====================
    
    private static class LoggingInterceptor implements Interceptor {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            
            LOGGER.fine("Request: " + request.method() + " " + request.url());
            
            long startTime = System.nanoTime();
            Response response = chain.proceed(request);
            long endTime = System.nanoTime();
            
            LOGGER.fine("Response: " + response.code() + " in " +
                    (endTime - startTime) / 1_000_000 + "ms");
            
            return response;
        }
    }
}

/*
 * EXAMPLE USAGE:
 * 
 * import com.authsystem.sdk.AuthClient;
 * 
 * public class Example {
 *     public static void main(String[] args) {
 *         // Initialize the client using builder pattern
 *         AuthClient client = new AuthClient.Builder()
 *                 .apiBaseUrl("https://www.licensify.space")
 *                 .apiKey("your-api-key-here")
 *                 .build();
 *         
 *         try {
 *             // Get the current device HWID
 *             String hwid = client.getHWID();
 *             System.out.println("Device HWID: " + hwid);
 *             
 *             // Register a new user
 *             client.register("johndoe", "SecurePassword123", "john@example.com", "LICENSE-KEY-123")
 *                     .thenAccept(response -> {
 *                         if (response.success) {
 *                             System.out.println("Registration successful!");
 *                         }
 *                     })
 *                     .exceptionally(throwable -> {
 *                         System.err.println("Registration failed: " + throwable.getMessage());
 *                         return null;
 *                     })
 *                     .get(); // Wait for completion
 *             
 *             // Login (first time - device will be locked)
 *             client.login("johndoe", "SecurePassword123")
 *                     .thenAccept(response -> {
 *                         if (response.success) {
 *                             System.out.println("Login successful! Welcome " + response.user.username);
 *                             
 *                             if (response.user.isFirstLogin) {
 *                                 System.out.println("This device has been locked to your account!");
 *                             }
 *                             
 *                             System.out.println("HWID Locked: " + response.user.hwidLocked);
 *                             System.out.println("Current IP: " + response.user.ip);
 *                         }
 *                     })
 *                     .exceptionally(throwable -> {
 *                         if (throwable.getCause() instanceof AuthClient.AuthException) {
 *                             AuthClient.AuthException authEx = (AuthClient.AuthException) throwable.getCause();
 *                             System.err.println("Auth error [" + authEx.getErrorCode() + "]: " + authEx.getMessage());
 *                             
 *                             // Handle specific errors
 *                             switch (authEx.getErrorCode()) {
 *                                 case "HWID_LOCKED":
 *                                     System.err.println("This account is locked to a different device!");
 *                                     break;
 *                                 case "INVALID_CREDENTIALS":
 *                                     System.err.println("Wrong username or password!");
 *                                     break;
 *                                 case "LICENSE_EXPIRED":
 *                                     System.err.println("Your license has expired!");
 *                                     break;
 *                             }
 *                         }
 *                         return null;
 *                     })
 *                     .get(); // Wait for completion
 *             
 *             // Validate license
 *             client.validateLicense("LICENSE-KEY-123")
 *                     .thenAccept(response -> {
 *                         System.out.println("License valid: " + response.valid);
 *                     })
 *                     .get();
 *             
 *             // Session is automatically maintained with heartbeat
 *             Thread.sleep(10000);
 *             
 *             // Check session
 *             client.checkSession()
 *                     .thenAccept(response -> {
 *                         System.out.println("Session valid: " + response.valid);
 *                     })
 *                     .get();
 *             
 *             // Logout
 *             client.logout()
 *                     .thenAccept(response -> {
 *                         System.out.println("Logout successful");
 *                     })
 *                     .get();
 *             
 *         } catch (Exception e) {
 *             e.printStackTrace();
 *         } finally {
 *             // Cleanup
 *             client.close();
 *         }
 *     }
 * }
 */

File: public/sdk/python/auth_client.py
"""
Python Authentication Client with HWID-based Device Binding
Cross-platform support for Windows, Linux, and macOS
"""

import asyncio
import hashlib
import platform
import subprocess
import uuid
import aiohttp
import logging
from typing import Optional, Dict, Any
from dataclasses import dataclass
from datetime import datetime, timedelta


class AuthException(Exception):
    """Custom exception for authentication errors"""
    def __init__(self, message: str, error_code: str):
        self.message = message
        self.error_code = error_code
        super().__init__(f"[{error_code}] {message}")


@dataclass
class UserInfo:
    """User information model"""
    id: str
    username: str
    email: str
    hwid: str
    hwid_locked: bool
    ip: str
    is_first_login: bool = False
    expires_at: Optional[str] = None


@dataclass
class LoginResponse:
    """Login response model"""
    success: bool
    message: str
    token: Optional[str] = None
    user: Optional[UserInfo] = None


class HWIDGenerator:
    """Cross-platform Hardware ID generator"""
    
    @staticmethod
    def generate() -> str:
        """
        Generate unique HWID based on the current platform
        Windows: User SID → System UUID → Processor ID
        Linux: Machine ID → DMI UUID → MAC Address
        macOS: Hardware UUID → MAC Address → Serial Number
        """
        system = platform.system()
        
        try:
            if system == "Windows":
                return HWIDGenerator._generate_windows()
            elif system == "Linux":
                return HWIDGenerator._generate_linux()
            elif system == "Darwin":  # macOS
                return HWIDGenerator._generate_macos()
            else:
                return HWIDGenerator._generate_fallback()
        except Exception as e:
            logging.warning(f"HWID generation failed: {e}, using fallback")
            return HWIDGenerator._generate_fallback()
    
    @staticmethod
    def _generate_windows() -> str:
        """Generate HWID for Windows"""
        try:
            # Primary: Windows User SID
            import win32security
            token = win32security.OpenProcessToken(
                win32security.GetCurrentProcess(),
                win32security.TOKEN_QUERY
            )
            sid = win32security.GetTokenInformation(
                token, win32security.TokenUser
            )[0]
            sid_string = win32security.ConvertSidToStringSid(sid)
            return HWIDGenerator._hash_string(sid_string)
        except:
            pass
        
        try:
            # Secondary: System UUID from WMI
            import wmi
            c = wmi.WMI()
            for item in c.Win32_ComputerSystemProduct():
                if item.UUID:
                    return HWIDGenerator._hash_string(item.UUID)
        except:
            pass
        
        # Tertiary: MAC Address
        mac = HWIDGenerator._get_mac_address()
        if mac:
            return HWIDGenerator._hash_string(mac)
        
        return HWIDGenerator._generate_fallback()
    
    @staticmethod
    def _generate_linux() -> str:
        """Generate HWID for Linux"""
        try:
            # Primary: Machine ID
            with open('/etc/machine-id', 'r') as f:
                machine_id = f.read().strip()
                if machine_id:
                    return HWIDGenerator._hash_string(machine_id)
        except:
            pass
        
        try:
            # Secondary: DMI System UUID
            with open('/sys/class/dmi/id/product_uuid', 'r') as f:
                dmi_uuid = f.read().strip()
                if dmi_uuid:
                    return HWIDGenerator._hash_string(dmi_uuid)
        except:
            pass
        
        # Tertiary: MAC Address
        mac = HWIDGenerator._get_mac_address()
        if mac:
            return HWIDGenerator._hash_string(mac)
        
        return HWIDGenerator._generate_fallback()
    
    @staticmethod
    def _generate_macos() -> str:
        """Generate HWID for macOS"""
        try:
            # Primary: Hardware UUID
            result = subprocess.run(
                ['system_profiler', 'SPHardwareDataType'],
                capture_output=True,
                text=True
            )
            for line in result.stdout.split('\n'):
                if 'Hardware UUID' in line:
                    hw_uuid = line.split(':')[1].strip()
                    return HWIDGenerator._hash_string(hw_uuid)
        except:
            pass
        
        # Secondary: MAC Address
        mac = HWIDGenerator._get_mac_address()
        if mac:
            return HWIDGenerator._hash_string(mac)
        
        return HWIDGenerator._generate_fallback()
    
    @staticmethod
    def _get_mac_address() -> Optional[str]:
        """Get the MAC address of the first physical network interface"""
        try:
            mac = ':'.join(['{:02x}'.format((uuid.getnode() >> elements) & 0xff)
                          for elements in range(0, 8*6, 8)][::-1])
            return mac if mac != '00:00:00:00:00:00' else None
        except:
            return None
    
    @staticmethod
    def _generate_fallback() -> str:
        """Fallback HWID generation"""
        fallback_string = f"{platform.node()}-{platform.machine()}-{uuid.getnode()}"
        return HWIDGenerator._hash_string(fallback_string)
    
    @staticmethod
    def _hash_string(input_str: str) -> str:
        """Hash a string using SHA-256"""
        return hashlib.sha256(input_str.encode('utf-8')).hexdigest()


class AuthClient:
    """
    Main authentication client with HWID-based device binding
    """
    
    def __init__(self, api_base_url: str, api_key: str):
        """
        Initialize the authentication client
        
        Args:
            api_base_url: Base URL of the authentication API
            api_key: Your application's API key
        """
        self.api_base_url = api_base_url.rstrip('/')
        self.api_key = api_key
        self._session: Optional[aiohttp.ClientSession] = None
        self._jwt_token: Optional[str] = None
        self._heartbeat_task: Optional[asyncio.Task] = None
        self._hwid: Optional[str] = None
        
        # Configure logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        self.logger = logging.getLogger('AuthClient')
    
    @property
    def hwid(self) -> str:
        """Get the current device HWID (cached)"""
        if not self._hwid:
            self._hwid = HWIDGenerator.generate()
        return self._hwid
    
    async def __aenter__(self):
        """Async context manager entry"""
        await self._ensure_session()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit"""
        await self.close()
    
    async def _ensure_session(self):
        """Ensure aiohttp session exists"""
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers={
                    'User-Agent': 'AuthSystem-Python-SDK/1.0',
                    'Content-Type': 'application/json'
                }
            )
    
    async def close(self):
        """Close the client and cleanup resources"""
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
            try:
                await self._heartbeat_task
            except asyncio.CancelledError:
                pass
        
        if self._session and not self._session.closed:
            await self._session.close()
    
    async def register(
        self,
        username: str,
        password: str,
        email: str,
        license_key: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Register a new user account
        
        Args:
            username: Desired username
            password: Account password
            email: User email address
            license_key: Optional license key
            
        Returns:
            Registration response dictionary
        """
        payload = {
            'apiKey': self.api_key,
            'username': username,
            'password': password,
            'email': email,
            'hwid': self.hwid
        }
        
        if license_key:
            payload['licenseKey'] = license_key
        
        return await self._send_request('POST', '/api/user/create', payload)
    
    async def login(self, username: str, password: str) -> LoginResponse:
        """
        Login with username and password
        The server will automatically bind this device's HWID to the account
        
        Args:
            username: Account username
            password: Account password
            
        Returns:
            LoginResponse object with user info and token
        """
        payload = {
            'apiKey': self.api_key,
            'username': username,
            'password': password,
            'hwid': self.hwid
        }
        
        response = await self._send_request('POST', '/api/user/login', payload)
        
        if response.get('success'):
            self._jwt_token = response.get('token')
            self._start_heartbeat()
            
            user_data = response.get('user', {})
            user_info = UserInfo(
                id=user_data.get('id'),
                username=user_data.get('username'),
                email=user_data.get('email'),
                hwid=user_data.get('hwid'),
                hwid_locked=user_data.get('hwidLocked', False),
                ip=user_data.get('ip'),
                is_first_login=user_data.get('isFirstLogin', False),
                expires_at=user_data.get('expiresAt')
            )
            
            return LoginResponse(
                success=True,
                message=response.get('message'),
                token=self._jwt_token,
                user=user_info
            )
        else:
            return LoginResponse(
                success=False,
                message=response.get('message', 'Login failed')
            )
    
    async def validate_license(self, license_key: str) -> Dict[str, Any]:
        """
        Validate a license key
        
        Args:
            license_key: License key to validate
            
        Returns:
            Validation response dictionary
        """
        payload = {
            'apiKey': self.api_key,
            'licenseKey': license_key,
            'hwid': self.hwid
        }
        
        return await self._send_request('POST', '/api/auth/validate', payload)
    
    async def check_session(self) -> Dict[str, Any]:
        """
        Check if the current session is still valid
        
        Returns:
            Session status dictionary
        """
        return await self._send_request('GET', '/api/user/session')
    
    async def logout(self) -> Dict[str, Any]:
        """
        Logout and invalidate the current session
        
        Returns:
            Logout response dictionary
        """
        self._stop_heartbeat()
        response = await self._send_request('POST', '/api/user/logout')
        self._jwt_token = None
        return response
    
    def _start_heartbeat(self):
        """Start the session heartbeat task"""
        if self._heartbeat_task is None or self._heartbeat_task.done():
            self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())
    
    def _stop_heartbeat(self):
        """Stop the session heartbeat task"""
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
            self._heartbeat_task = None
    
    async def _heartbeat_loop(self):
        """Background task to keep session alive"""
        try:
            while True:
                await asyncio.sleep(300)  # 5 minutes
                try:
                    await self.check_session()
                    self.logger.debug("Heartbeat successful")
                except Exception as e:
                    self.logger.warning(f"Heartbeat failed: {e}")
        except asyncio.CancelledError:
            self.logger.debug("Heartbeat cancelled")
    
    async def _send_request(
        self,
        method: str,
        endpoint: str,
        payload: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Send HTTP request to the API
        
        Args:
            method: HTTP method (GET, POST, etc.)
            endpoint: API endpoint path
            payload: Optional request payload
            
        Returns:
            Response dictionary
        """
        await self._ensure_session()
        
        url = f"{self.api_base_url}{endpoint}"
        headers = {
            'X-HWID': self.hwid
        }
        
        if self._jwt_token:
            headers['Authorization'] = f'Bearer {self._jwt_token}'
        
        try:
            async with self._session.request(
                method,
                url,
                json=payload if payload else None,
                headers=headers
            ) as response:
                response_data = await response.json()
                
                if response.status >= 400:
                    error_code = response_data.get('errorCode', 'REQUEST_FAILED')
                    error_message = response_data.get('message', 'Request failed')
                    raise AuthException(error_message, error_code)
                
                return response_data
                
        except aiohttp.ClientError as e:
            raise AuthException(f"Network error: {str(e)}", "NETWORK_ERROR")
        except AuthException:
            raise
        except Exception as e:
            raise AuthException(f"Unexpected error: {str(e)}", "UNKNOWN_ERROR")


# Example Usage
async def example_usage():
    """
    Example demonstrating how to use the AuthClient
    """
    # Initialize the client
    async with AuthClient(
        "https://www.licensify.space",
        "your-api-key-here"
    ) as client:
        
        try:
            # Get the current device HWID
            hwid = client.hwid
            print(f"Device HWID: {hwid}")
            
            # Register a new user
            register_result = await client.register(
                username="johndoe",
                password="SecurePassword123",
                email="john@example.com",
                license_key="LICENSE-KEY-123"
            )
            print(f"Registration: {register_result}")
            
            # Login (first time - device will be locked)
            login_result = await client.login("johndoe", "SecurePassword123")
            
            if login_result.success:
                print(f"Login successful! Welcome {login_result.user.username}")
                
                if login_result.user.is_first_login:
                    print("This device has been locked to your account!")
                
                print(f"HWID Locked: {login_result.user.hwid_locked}")
                print(f"Current IP: {login_result.user.ip}")
                
                # Validate license
                validate_result = await client.validate_license("LICENSE-KEY-123")
                print(f"License validation: {validate_result}")
                
                # Session is automatically maintained with heartbeat
                await asyncio.sleep(10)
                
                # Check session
                session_status = await client.check_session()
                print(f"Session status: {session_status}")
                
                # Logout
                logout_result = await client.logout()
                print(f"Logout: {logout_result}")
            
        except AuthException as e:
            print(f"Authentication error [{e.error_code}]: {e.message}")
            
            # Handle specific errors
            if e.error_code == "HWID_LOCKED":
                print("This account is locked to a different device!")
            elif e.error_code == "INVALID_CREDENTIALS":
                print("Wrong username or password!")
            elif e.error_code == "LICENSE_EXPIRED":
                print("Your license has expired!")


if __name__ == "__main__":
    # Run the example
    asyncio.run(example_usage())

